### PR TITLE
provider: Introduce Pod Security Standard for non-CC and CC based Clusters

### PR DESCRIPTION
**Pull request by chrischdi**
### What this PR does / why we need it

Introduces:
* Overlay for ClusterClasses to enable and configure Pod Security Standard for v1.23 clusters
* Overlay for non-ClusterClass clusters to enable and configure Pod Security Standard for v1.23 clusters
* Migration code for non-ClusterClass clusters to enable Pod Security Standard when upgrading to v1.23

Follow-up PR: raise the bar to v1.24 clusters instead of v1.23 clusters, as soon as v1.24 is available in tanzu-framework.

### Which issue(s) this PR fixes

Fixes #

### Release note

```release-note
Enable pod security admission per default for Cluster in audit and warn mode. Also introduces variables to the ClusterClasses to make it configurable for ClusterClass based clusters.
```

### CODE REVIEW

File: packages/tkg-clusterclass/hack/sync-cc.sh
Comment (chrischdi) [inline packages/tkg-clusterclass/hack/sync-cc.sh:20]: Did not find a current approach to have a directory for overlays which are shared across ClusterClasses of different providers.

File: tkg/client/upgrade_cluster.go
Comment (chrischdi) [inline tkg/client/upgrade_cluster.go:1149]: I hope this integration to enable it for v1.24 is as intended in tanzu? Happy to refactor if this is the wrong place to implement the functionality.

File: packages/tkg-clusterclass-aws/bundle/config/upstream/overlay-podsecuritystandard.yaml
Comment (zjs) [inline packages/tkg-clusterclass-aws/bundle/config/upstream/overlay-podsecuritystandard.yaml:7]: Naive question: is there some way to put this under [tkg-clusterclass](https://github.com/vmware-tanzu/tanzu-framework/tree/main/packages/tkg-clusterclass) to avoid needing to duplicate this info across each of the infrastructure providers?

File: packages/tkg-clusterclass-aws/bundle/config/upstream/overlay-podsecuritystandard.yaml
Comment (zjs) [inline packages/tkg-clusterclass-aws/bundle/config/upstream/overlay-podsecuritystandard.yaml:22]: Under inclusive naming guidelines, I think we prefer to use deactivated/activated instead of disabled/enabled, for any code we control.

File: packages/tkg-clusterclass-vsphere/bundle/config/upstream/overlay-podsecuritystandard.yaml
Comment (zjs) [inline packages/tkg-clusterclass-vsphere/bundle/config/upstream/overlay-podsecuritystandard.yaml:89]: "capz-system" should be replaced with something else for vSphere, I think?

File: packages/tkg-clusterclass-docker/bundle/config/upstream/overlay-podsecuritystandard.yaml
Comment (zjs) [inline packages/tkg-clusterclass-docker/bundle/config/upstream/overlay-podsecuritystandard.yaml:89]: "capz-system" should be replaced with something else for Docker, I think?

File: packages/tkg-clusterclass-aws/bundle/config/upstream/overlay-podsecuritystandard.yaml
Comment (zjs) [inline packages/tkg-clusterclass-aws/bundle/config/upstream/overlay-podsecuritystandard.yaml:89]: "capz-system" should be replaced with something else for AWS, I think?

File: providers/ytt/03_customizations/pod_security_admission.yaml
Comment (zjs) [inline providers/ytt/03_customizations/pod_security_admission.yaml:22]: I think we'll want to pin to a specific version, instead of just using `latest`.

**Review by zjs — COMMENTED** [review]
Took a quick look. I'm not an expert on any of this; hopefully someone who knows more can take a closer look. Thanks for your help with this!

File: packages/tkg-clusterclass-aws/bundle/config/upstream/overlay-podsecuritystandard.yaml
Comment (chrischdi) [reply packages/tkg-clusterclass-aws/bundle/config/upstream/overlay-podsecuritystandard.yaml]: resolved, we only have to set it for Azure, but now it will also only get set for azure :-)

File: packages/tkg-clusterclass-docker/bundle/config/upstream/overlay-podsecuritystandard.yaml
Comment (chrischdi) [reply packages/tkg-clusterclass-docker/bundle/config/upstream/overlay-podsecuritystandard.yaml]: resolved, we only have to set it for Azure, but now it will also only get set for azure :-)

File: packages/tkg-clusterclass-vsphere/bundle/config/upstream/overlay-podsecuritystandard.yaml
Comment (chrischdi) [reply packages/tkg-clusterclass-vsphere/bundle/config/upstream/overlay-podsecuritystandard.yaml]: resolved, we only have to set it for Azure, but now it will also only get set for azure :-)

File: packages/tkg-clusterclass-aws/bundle/config/upstream/overlay-podsecuritystandard.yaml
Comment (chrischdi) [reply packages/tkg-clusterclass-aws/bundle/config/upstream/overlay-podsecuritystandard.yaml]: Great point 👍  I refactored it to deactivated.

File: packages/tkg-clusterclass-aws/bundle/config/upstream/overlay-podsecuritystandard.yaml
Comment (chrischdi) [reply packages/tkg-clusterclass-aws/bundle/config/upstream/overlay-podsecuritystandard.yaml]: Great idea, wasn't aware that this is possible.

Moved to a separate secret via `tkg-clusterclass` which is used as overlay for the provider specific packageinstall.

File: providers/ytt/03_customizations/pod_security_admission.yaml
Comment (chrischdi) [reply providers/ytt/03_customizations/pod_security_admission.yaml:22]: TODO by me. I need to take another look at it.

File: packages/tkg-clusterclass/bundle/config/packageinstalls/tkg-infra-clusterclass-pi.yaml
Comment (tenczar) [inline packages/tkg-clusterclass/bundle/config/packageinstalls/tkg-infra-clusterclass-pi.yaml:151]: nit: typo
```suggestion
                description: "exemption configuration for the PodSecurityConfiguration."
```

File: packages/tkg-clusterclass/bundle/config/packageinstalls/tkg-infra-clusterclass-pi.yaml
Comment (tenczar) [inline packages/tkg-clusterclass/bundle/config/packageinstalls/tkg-infra-clusterclass-pi.yaml:210]: Same comment about latest. Either TODO, or document the reason for using latest.

File: packages/tkg-clusterclass/bundle/config/packageinstalls/tkg-infra-clusterclass-pi.yaml
Comment (tenczar) [inline packages/tkg-clusterclass/bundle/config/packageinstalls/tkg-infra-clusterclass-pi.yaml:207]: I might recommend something like the following
```suggestion
                        enforce: "{{ if .podSecurityStandard.enforce -}}
                            {{  .podSecurityStandard.enforce }}
                         {{- end }}"
```

For some of these it might make it easier to parse. You will need to use the proper `{{-` or `-}}` to make sure it is formatted correctly.

**Review by tenczar — APPROVED** [review]
LGTM

**Comment by chrischdi** [comment]
Thanks for the reviews. I adressed the comments and rebased.

Is there something like

/hold

here? I think it would be awesome to maybe also get a review from @frapposelli and/or @randomvariable post PTO/Kubecon 😊

File: packages/tkg-clusterclass-aws/bundle/config/upstream/overlay-podsecuritystandard.yaml
Comment (chrischdi) [reply packages/tkg-clusterclass-aws/bundle/config/upstream/overlay-podsecuritystandard.yaml]: @zjs question: I now moved the overlay to the `tkg-clusterclass` package.

However: when creating a management cluster the current code seems to **not** use the package `tkg-clusterclass` package (at least it does not for `--dry-run` and at the current code state).

I want to confirm on the behaviour as soon as the CAP* Provider PRs are in to verify it.

**Comment by alfredthenarwhal** [comment]
> @chrischdi: /test install-aws

Tests can be triggered only by the trusted reviewers (users mentioned in CODEOWNERS file).

**Comment by chrischdi** [comment]
> Cluster Generation A/B Results: https://storage.googleapis.com/tkg-clustergen/3693/20221102152209/clustergen.diff.txt Author/reviewers: Please review to verify that the effects on the generated cluster configurations are exactly what the PR intended, and give a thumbs-up if so.

The intended new defaults apply to the clusters 👍

File: packages/tkg-clusterclass-aws/bundle/config/upstream/overlay-podsecuritystandard.yaml
Comment (chrischdi) [reply packages/tkg-clusterclass-aws/bundle/config/upstream/overlay-podsecuritystandard.yaml]: Correction: when integrating overlays into `tkg-clusterclass`:
* `tanzu mc create --dryrun` won't include the overlays
* `tanzu mc create --dryrun=false` will use the packageinstall and the overlays will apply.

File: packages/tkg-clusterclass/bundle/config/packageinstalls/tkg-infra-clusterclass-pi.yaml
Comment (tenczar) [inline packages/tkg-clusterclass/bundle/config/packageinstalls/tkg-infra-clusterclass-pi.yaml:88]: will this cause an error somewhere?
```suggestion
  kcp-empty-array-overlay.yaml: |
```

File: packages/tkg-clusterclass/bundle/config/packageinstalls/tkg-infra-clusterclass-pi.yaml
Comment (tenczar) [inline packages/tkg-clusterclass/bundle/config/packageinstalls/tkg-infra-clusterclass-pi.yaml:201]: has this been fixed? I remember having issue early on where adding with `-` wasn't properly adding to the end of the array. I just want to confirm that this has been tested to work.

**Review by tenczar — APPROVED** [review]
LGTM

File: packages/tkg-clusterclass/bundle/config/packageinstalls/tkg-infra-clusterclass-pi.yaml
Comment (chrischdi) [reply packages/tkg-clusterclass/bundle/config/packageinstalls/tkg-infra-clusterclass-pi.yaml:88]: Nope, renaming is fine. Carvel will use all files of the config map 👍

File: packages/tkg-clusterclass/bundle/config/packageinstalls/tkg-infra-clusterclass-pi.yaml
Comment (chrischdi) [reply packages/tkg-clusterclass/bundle/config/packageinstalls/tkg-infra-clusterclass-pi.yaml:201]: I tested it if it works, yes.

Adding with `-` requires to have an empty array first. This gets added by the `kcp-empty-array-overlay.yaml`.
Note: having the empty array in the to-be patched object (in this case `KubeadmControlPlaneTemplate`) does not always work because the field may be treated with omitEmpty. Thats why it is required to have the "add empty array" patch.

With Halifax and Runtime SDK we will be able to write better patches using the topology mutation hooks.

**Comment by chrischdi** [comment]
Closing in favor of #3862 to also run ClusterClass integration tests.

**Pull request closed without merging**
